contracts/implementations/MultipleMinter.sol

```diff
@@ -279,4 +279,140 @@ contract MultipleMinter is
+        );
+
+        if (
+            !Eip712CheckerInternal._verifySignature(
+                data.syntheticDeviceAddr,
+                message,
+                data.syntheticDeviceSig
+            )
+        ) revert InvalidSdSignature();
+
+        // ----- START Vehicle mint -----
+        uint256 newTokenIdVehicle = INFT(vehicleIdProxyAddress).safeMint(
+            data.owner
+        );
+
+        emit VehicleNodeMintedWithDeviceDefinition(
+            data.manufacturerNode,
+            newTokenIdVehicle,
+            data.owner,
+            data.deviceDefinitionId
+        );
+
+        (bytes32 attributesHash, bytes32 infosHash) = _setInfosHash(
+            newTokenIdVehicle,
+            data.attrInfoPairsVehicle
+        );
+
+        message = keccak256(
+            abi.encode(
+                MINT_VEHICLE_WITH_DD_TYPEHASH,
+                data.manufacturerNode,
+                data.owner,
+                keccak256(bytes(data.deviceDefinitionId)),
+                attributesHash,
+                infosHash
+            )
+        );
+
+        if (
+            !Eip712CheckerInternal._verifySignature(
+                data.owner,
+                message,
+                data.vehicleOwnerSig
+            )
+        ) revert InvalidOwnerSignature();
+        // ----- END Vehicle mint -----
+
+        // ----- START Synthetic Device mint and attributes -----
+        uint256 newTokenIdDevice = INFT(sdIdProxyAddress).safeMint(data.owner);
+
+        emit SyntheticDeviceNodeMinted(
+            data.integrationNode,
+            newTokenIdDevice,
+            newTokenIdVehicle,
+            data.syntheticDeviceAddr,
+            data.owner
+        );
+
+        if (data.attrInfoPairsDevice.length > 0)
+            _setInfos(newTokenIdDevice, data.attrInfoPairsDevice);
+        // ----- END Synthetic Device mint and attributes -----
+
+        // ----- Internal contract state change -----
+        NodesStorage
+        .getStorage()
+        .nodes[vehicleIdProxyAddress][newTokenIdVehicle].parentNode = data
+            .manufacturerNode;
+        vs.vehicleIdToDeviceDefinitionId[newTokenIdVehicle] = data
+            .deviceDefinitionId;
+
+        NodesStorage
+        .getStorage()
+        .nodes[sdIdProxyAddress][newTokenIdDevice].parentNode = data
+            .integrationNode;
+
+        MapperStorage.getStorage().nodeLinks[vehicleIdProxyAddress][
+            sdIdProxyAddress
+        ][newTokenIdVehicle] = newTokenIdDevice;
+        MapperStorage.getStorage().nodeLinks[sdIdProxyAddress][
+            vehicleIdProxyAddress
+        ][newTokenIdDevice] = newTokenIdVehicle;
+
+        sds.deviceAddressToNodeId[data.syntheticDeviceAddr] = newTokenIdDevice;
+        sds.nodeIdToDeviceAddress[newTokenIdDevice] = data.syntheticDeviceAddr;
+
+        ChargingInternal._chargeDcx(msg.sender, MINT_VEHICLE_OPERATION);
+
+        INFT(vehicleIdProxyAddress).setSacd(newTokenIdVehicle, sacdInput);
```

Comment:
Will the DCX charge be reverted if setSACD fails?

Reply:
Yes. If anything fails, everything reverts

Reply:
The word "transaction" is apt.